social_core/backends/microsoft.py

```diff
@@ -45,8 +45,11 @@ def get_user_id(self, details, response):
 
     def get_user_details(self, response):
         """Return user details from Microsoft online account"""
+        email = response.get('mail')
+        if not email and '@' in response.get('userPrincipalName'):
+            email = response.get('userPrincipalName')
         return {'username': response.get('displayName', ''),
```

Comment:
* `displayName` is the user's name, e.g., John Lennon. Shouldn't the username property be set to a normal one like `johnlennon` instead? If yes, we can just use the UPN because it's also the username for Microsoft.

Reply:
Looking at the docs, that looks exactly what's needed here, the `UPN` should be used as the `username`, if it contins an `@` on it, then the prefix needs to be extracted and used instead.